default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
+    blackhole = len(AIstate.empireStars.get(fo.starType.blackHole, [])) != 0
+    neutron = len(AIstate.empireStars.get(fo.starType.neutron , [])) != 0
+
+    hull = 1
+    offtrack_hull = 0.05
+    offtrack_subhull = 0.25
+    # select one hull and specialize it, but some AI may want to have more hulls, by random
+    org = hull if index & 1 == 0 or (index + 97) % 227 < 21 else offtrack_hull
+    robotic = hull if index & 1 == 1 or (index + 107) % 229 < 22 else offtrack_hull
+    if ColonisationAI.got_ast:
+        extra = (index + 113) % 223 < 11
+        asteroid = hull if index & 2 == 0 or extra else offtrack_hull
+        if asteroid == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+    else:
+        asteroid = offtrack_hull
+    if bright_star or blackhole:
+        extra = (index + 103) % 211 < 10
+        energy = hull if index & 4 == 0 or extra else offtrack_hull
+        if energy == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+            asteroid = offtrack_hull
+    else:
+        energy = offtrack_hull
+
+    # further specialization for robotic hulls
+    contgrav = hull if robotic == hull and index & 8 == 0 else offtrack_subhull
+    nanorobo = hull if robotic == hull and index & 8 == 1 else offtrack_subhull
+    flux = hull if robotic == hull and (index + 17) % 43 < 22 else offtrack_subhull
+    # further specialization for asteroid hulls
+    astheavy = hull if asteroid == hull and index % 3 == 0 else offtrack_subhull
+    astswarm = hull if asteroid == hull and index % 3 == 1 else offtrack_subhull
+    astcamo = hull if asteroid == hull and index % 3 == 2 else offtrack_subhull
+    # further specialization for organic hulls
+    orgneural = hull if org == hull and index & 8 == 0 else offtrack_subhull
+    orgraven = hull if org == hull and index & 8 == 1 else offtrack_subhull
+    orgendosym = hull if org == hull and (index + 17) % 43 < 22 else offtrack_subhull
+    # further specialization for energy hulls
+    energyfrac = hull if energy == hull and index & 8 == 0 else offtrack_subhull
+    energymag = hull if energy == hull and index & 8 == 1 else offtrack_subhull
+    # repair techs may be skipped if AI decides to go for nanorobotic hull which full-repairs
+    repair = 1 if not nanorobo == hull or (index + 37) % 103 < 77 else 0.3
+
+    # AI may skip weapon lines
+    weapon = 1
+    massdriver = weapon if index % 47 < 36 else 0
+    laser = weapon if (index % 83 < 36 if massdriver == weapon else index % 47 < 41) else 0
+    plasmacannon = weapon if (index % 83 < 36 if laser == weapon else index % 47 < 41) else 0
+    deathray = weapon
+
+    armor = 1
+
+    engine = 1 if ((index + 13) % 31 < 26 if ColonisationAI.galaxy_is_sparse() else (index + 13) % 31 < 5) else 0
+    fuel = 1 if ((index + 17) % 37 < 31 if ColonisationAI.galaxy_is_sparse() else (index + 17) % 37 < 3) else 0
```

Comment:
In games that I play, I rarely need to use engines/fuels because I tend to reduce number of front line systems and send any ships to front line upon completion. I also tends to take systems slowly (i.e. turtling) and hence engines/fuels are not actually important for my playstyle

So I presume it is more a personal choice lol :smile: